added fluentd-elasticsearch

monotek opened the pull request:
This pr adds a fluentd chart wich exports to elasticsearch using the kubernetes metadata and googles dectect-exeptions plugin to have proper multiline logging for java stacktraces and so on.

Currently i use an own fluentd 1.1.0 dockerimage from: https://hub.docker.com/r/monotek/docker-fluentd-elasticsearch/

I'll added another pull request (https://github.com/kubernetes/kubernetes/pull/58525) to the Kubernetes repo to update the original image in: https://console.cloud.google.com/gcr/images/google-containers/GLOBAL/fluentd-elasticsearch

It already got the detect-exeptions plugin with: https://github.com/kubernetes/kubernetes/pull/58063 so it only needs the fluentd 1.1.0 update.

Commit:
added fluentd-elasticsearch chart to icubator

Commit:
fixed configmap for fluentd 1.1.0

monotek commented:
/assign @viglesiasce

monotek committed:
Update values.yaml

monotek committed:
Update Chart.yaml

monotek committed:
Update README.md

monotek committed:
added fluentd custom config option

monotek committed:
disable fluentd custom config by default

monotek commented:
https://github.com/kubernetes/kubernetes/pull/58525 has been merged, so chart will work with official kubernetes fluentd elasticsearch docker image as soon as new image is build.

viglesiasce commented:
/ok-to-test

viglesiasce commented:
@monotek let me know when the image is available

rendhalver commented:
I have put in a chart that does something similar here #3560 
Does your setup include the kubernetes metadata plugin?
Last I looked that didn't work with fluentd 1.x.

rendhalver commented:
I guess it does include the kubernetes metadata.
There was also some discussion on my original PR for this to make the output more configurable so it worked with more that just elasticsearch.
Is that possible with your setup?

rendhalver commented:
Thanks for this.
@monotek we set this up locally and it works well.
The only problem we found was that it didn't add a pod to our three master nodes so any logs there were not showing up.
I have some code to add that in if you would like it.
I can send you a PR to your branch or just open it up after this gets merged. Which would you prefer?

Commit:
Add tolerations for DaemonSet

Commit:
Add annotation to restart DaemonSet when ConfigMap changes

monotek commented:
Thanks for testing and contributing :-)
Maybe adding it in one PR is better for having it implemented faster.
So i would kindly ask you to add a PR against my repo. 
I'll update this PR...

About the other outputs... 
This would be possible but i guess we would have to change the name of the chart to something more generic and we would also need a new image as the current one only supports elasticsearch. So maybe keep this for now for a later PR or even better lets work on a new generic fluentd chart.

rendhalver commented:
I am working on extra fixes as well.
We are working on splitting the logs for each namespace into their own index so we need the output.conf configurable and the ability to add an elasticsearch template.
Would that be helpful as well?

monotek commented:
I think so. 
Just other output plugins like splunk, mongo, kafka and so on would be out of scope...

rendhalver commented:
Those would need extra plugins so new containers so definitely out of scope.

viglesiasce commented:
/retest

Commit:
Move output.conf into variable so it can be configured from values

Add tolerations for DaemonSet
Add annotation to restart DaemonSet on ConfigMap changes

rendhalver commented:
@monotek I am trying to send you a PR with that extra code but I can't see your fork in in the list.
Not sure what's going on there. Any ideas?

monotek commented:
maybe you have to fork my branch? 
https://github.com/monotek/charts/tree/fluentd-elasticsearch

rendhalver commented:
@monotek This is my branch if that helps: 
https://github.com/powerhome/kubernetes-charts/tree/fluentd-elasticsearch-enhancements

monotek committed:
Merge pull request #1 from powerhome/fluentd-elasticsearch-enhancements

Fluentd elasticsearch enhancements

monotek commented:
@rendhalver
i've created a pull request to my own repo & merged your changes: https://github.com/monotek/charts/pull/1

rendhalver commented:
Thanks @monotek.
It looks like it's causing the build to fail. :( I will see if I can work out why.

rendhalver commented:
Ahh I think it's because it's not finding the images.

monotek commented:
Yep. I think because of:

https://github.com/kubernetes/kubernetes/pull/59665 & https://github.com/kubernetes/kubernetes/pull/58525

rendhalver commented:
Ah that would explain it.

monotek commented:
The image is available now :-)
https://console.cloud.google.com/gcr/images/google-containers/GLOBAL/fluentd-elasticsearch@sha256:b8c94527b489fb61d3d81ce5ad7f3ddbb7be71e9620a3a36e2bede2f2e487d73/details/info

rendhalver commented:
@monotek I am going to be writing a fluentd chart that uses a Deployment instead of as a DaemonSet so we can use it instead of Logstash in our Elastic-Stack clusters. Would your container image work in that capacity?

monotek commented:
Deployment makes no sense to me as you want to have fluentd on every node.

rendhalver commented:
It a different setup to what we are doing here.
We still have a bunch of servers outside Kubernetes that send logs to an elasticsearch cluster in kubernetes. Logstash isn't coping as well as we would like so we want to setup fluentd as a replacement.

monotek commented:
/retest

viglesiasce commented:
Thanks @monotek!

/lgtm
